test/model/message_list_test.dart

```diff
@@ -985,6 +1022,45 @@ void main() {
         checkNotifiedOnce();
       }));
 
+      test('fetchOlder backoff start, _reset, fetchOlder backoff ends, move fetch finishes', () => awaitFakeAsync((async) async {
+        await prepareNarrow(narrow, initialMessages);
+
+        connection.prepare(httpStatus: 400, json: {
+          'result': 'error', 'code': 'BAD_REQUEST', 'msg': 'Bad request'});
+        await check(model.fetchOlder()).throws<ZulipApiException>();
+        final backoffTimer = async.pendingTimers.single;
+        check(model).fetchOlderCoolingDown.isTrue();
+        checkHasMessages(initialMessages);
+        checkNotified(count: 2);
+
+        connection.prepare(delay: const Duration(seconds: 2), json: newestResult(
+          foundOldest: false,
+          messages: initialMessages + movedMessages,
+        ).toJson());
+        check(model).fetched.isTrue();
+        await store.handleEvent(eg.updateMessageEventMoveTo(
+          origTopic: movedMessages[0].topic,
+          origStreamId: otherStream.streamId,
+          newMessages: movedMessages,
+        ));
+        // check that _reset was caleed
+        check(model).fetched.isFalse();
+        check(model).fetchOlderCoolingDown.isFalse();
+        check(async.pendingTimers).contains(backoffTimer);
+        checkHasMessages([]);
+        checkNotifiedOnce();
+
+        // The first backoff is expected to be short enough to complete.
+        async.elapse(const Duration(seconds: 1));
+        check(model).fetchOlderCoolingDown.isFalse();
+        check(async.pendingTimers).not((x) => x.contains(backoffTimer));
+        checkNotNotified();
```

Comment:
When writing a test case, ideally we think through what scenarios would really stress the logic, and in particular would show clearly buggy symptoms if we had various plausible bugs.

So for example in the existing test just above this one, we let the fetch complete, and then check the model has no messages in its list:
```dart
        await fetchFuture;
        checkHasMessages([]);
        checkNotNotified();
```
(and then call `checkNotNotified`, mainly to ensure that `checkInvariants` gets called). For that test case that's good because if the `fetchOlder` call finished and then carried on, forgetting to check `generation`, then the symptom would be that it puts the fetched messages into the list.

Here, the current revision of this test checks that `fetchOlderCoolingDown` is still false. But that's not very informative, because although false is the value that flag had before this step, it's also exactly what the code under test would set the flag to if it *did* suffer from the race.

So to make a version of this test that really stresses the logic as it should, we should arrange a scenario where by the time the backoff completes, the flag has been set to true. Then we get to check here that it's still true — which confirms we don't have the bug where the code cheerfully sets it to false heedless of the new generation.

Reply:
Made a new test, "fetchOlder backoff A starts, _reset, move fetch finishes, fetchOlder backoff B starts, fetchOlder backoff A ends", for this. This also needed a refactor to make it easier to test with `BackoffMachine`, by offering a way to override the wait duration.